strawberry/fastapi/__init__.py

```diff
@@ -0,0 +1,188 @@
+import json
+from datetime import timedelta
+from typing import Any, Optional
+
+from fastapi import (
+    APIRouter,
+    BackgroundTasks,
+    Depends,
+    Request,
+    Response,
+    WebSocket,
+    status,
+)
+from fastapi.responses import HTMLResponse, JSONResponse, PlainTextResponse
+from strawberry.asgi.utils import get_graphiql_html
+from strawberry.exceptions import MissingQueryError
+from strawberry.fastapi.handlers.graphql_transport_ws_handler import (
+    GraphQLTransportWSHandler,
+)
+from strawberry.fastapi.handlers.graphql_ws_handler import GraphQLWSHandler
+from strawberry.file_uploads.utils import replace_placeholders_with_files
+from strawberry.http import GraphQLHTTPResponse, parse_request_data, process_result
+from strawberry.schema import BaseSchema
+from strawberry.subscriptions import GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL
+from strawberry.types import ExecutionResult
+from strawberry.utils.debug import pretty_print_graphql_operation
+
+
+async def get_root_value():
+    return None
+
+
+async def get_context(
+    background_tasks: BackgroundTasks, request: Request = None, ws: WebSocket = None
+) -> Optional[Any]:
+    return {"request": request or ws, "background_tasks": background_tasks}
+
+
+class GraphQL(APIRouter):
+    graphql_ws_handler_class = GraphQLWSHandler
+    graphql_transport_ws_handler_class = GraphQLTransportWSHandler
+
+    def __init__(
+        self,
+        schema: BaseSchema,
+        prefix: str = "",
+        graphiql: bool = True,
+        keep_alive: bool = False,
+        keep_alive_interval: float = 1,
+        debug: bool = False,
+        root_value_getter=get_root_value,
+        context_getter=get_context,
+        subscription_protocols=(GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL),
+        connection_init_wait_timeout: timedelta = timedelta(minutes=1),
+    ):
+        super().__init__(prefix=prefix)
+        self.schema = schema
+        self.graphiql = graphiql
+        self.keep_alive = keep_alive
+        self.keep_alive_interval = keep_alive_interval
+        self.debug = debug
+        self.root_value_getter = root_value_getter
+        self.context_getter = context_getter
+        self.protocols = subscription_protocols
+        self.connection_init_wait_timeout = connection_init_wait_timeout
+
+        @self.get(
+            "",
```

Comment:
Can this be set via Router? Like `graphiql_url`? And be analogous to `docs_url` and `redoc_url` FastAPI has?

In any case, this cannot be an empty string, can it? There's an issue when both the prefix and the endpoint have empty string:
```
Exception: Prefix and path cannot be both empty (path operation: home)
```
I mean, you can only see this error if you don't change the prefix. But I assume, as a user, that the default will work.

Reply:
Unless `static/graphiql.html` is changed to support custom fetch URL graphql and graphiql endpoints must be same. I'm not sure if this is standart in other GraphQL frameworks too but other extensions use same endpoint so I would prefer not to change this.